adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m

```diff
@@ -13,47 +13,92 @@
 // limitations under the License.
 
 #import "GADMediationAdapterPangle.h"
-
 #import "GADMediationAdapterPangleConstants.h"
+#import <BUAdSDK/BUAdSDK.h>
+#import "GADPangleRTBBannerRenderer.h"
+#import "GADPangleRTBInterstitialRenderer.h"
+#import "GADPangleRTBRewardedRenderer.h"
+#import "GADMediationAdapterPangleConstants.h"
+#import "GADMAdapterPangleUtils.h"
 
-@implementation GADMediationAdapterPangle
+static NSInteger _coppa = -1,_gdpr = -1, _ccpa = -1;
+
+@implementation GADMediationAdapterPangle {
+    /// Pangle banner ad wrapper.
+    GADPangleRTBBannerRenderer *_bannerRenderer;
+    /// Pangle interstitial ad wrapper.
+    GADPangleRTBInterstitialRenderer *_interstitialRenderer;
+    /// Pangle rewarded ad wrapper.
+    GADPangleRTBRewardedRenderer *_rewardedlRenderer;
+}
 
 - (void)collectSignalsForRequestParameters:(nonnull GADRTBRequestParameters *)params
                          completionHandler:(nonnull GADRTBSignalCompletionHandler)completionHandler {
-  // TODO: Start signal generation and call completionHandler when the signal generation finishes or fails.
-  completionHandler("", nil);
+    NSString *signals = [BUAdSDKManager getBiddingToken:nil];
+    if (completionHandler) {
+        completionHandler(signals,nil);
+    }
 }
 
 + (void)setUpWithConfiguration:(nonnull GADMediationServerConfiguration *)configuration
              completionHandler:(nonnull GADMediationAdapterSetUpCompletionBlock)completionHandler {
-  // TODO: Set up Pangle ad network SDK and perform any necessary prefetching or configuration work.
-  completionHandler(nil);
+    NSMutableSet *appIds = [[NSMutableSet alloc]init];
+    for (GADMediationCredentials *cred in configuration.credentials) {
+        NSString *appId = cred.settings[GADMAdapterPangleAppID];
+        if (appId.length) {
+            [appIds addObject:appId];
+        }
+    }
+    
+    if (appIds.count < 1) {
+        NSError *error = GADMAdapterPangleErrorWithCodeAndDescription(GADPangleErrorInvalidServerParameters, @"Pangle mediation configurations did not contain a valid App ID.");
+        completionHandler(error);
+        return;
+    }
+    
+    NSString *appID = [appIds anyObject];
+    if (appIds.count > 1) {
+        PangleLog(@"Found the following App IDs:%@. Please remove any app IDs which you are not using", appIds);
+        PangleLog(@"Configuring Pangle SDK with the app ID:%@", appID);
+    }
+    
+    BUAdSDKConfiguration *cog = [BUAdSDKConfiguration configuration];
+    cog.territory = BUAdSDKTerritory_NO_CN;
```

Comment:
What does this do?

Reply:
Pangle is oversea SDK，we have other AND in China